tests/source/imports/imports_granularity_default-with-dups.rs

```diff
@@ -0,0 +1,6 @@
+use crate::lexer;
```

Comment:
Could you add some test files that have multiple groups, and perhaps some that have comments too? Maybe something like this:

```rust
use crate::lexer::{tokens::TokenData /* foo */ };
use crate::lexer::{tokens::TokenData};
```

It might also be useful to have some duplicates but with varying visibility levels (we may not handle this well with the other variants either FWIW, but would be good to check while we're working on this one)

Reply:
> Could you add some test files that have multiple groups, and perhaps some that have comments too?

I didn't add such test files as there are some issues, especially related to comments, so maybe it is better to discuss first (here or in a new issue) which of these issues should be resolved, whether there are cases where the original code should be used, what should be the expected results, etc.

For example, the for the following input:
```rust
use crate::foo;
use crate::foo::self;
use crate::foo::self /* foo 1 */;
use crate::foo::{self /* foo 2 */ };
use crate::foo::{tokens::TokenData};
use crate::foo::{tokens::TokenData /* foo 3 */ };
use crate::foo::{self, tokens::TokenData /* foo 4 */};
use crate::bar::self /* bar 1 */;
use crate::bar::self;
use crate::foobar::self; /* foobar 1 */
use crate::foobar::self;
```

**"Crate"** granularity produce the following output:
```rust
use crate::bar; /* bar 1 */;
use crate::foo; /* foo 1 */;
use crate::foobar; /* foobar 1 */
use crate::{
    bar, foo,
    foo::{self /* foo 2 */, tokens::TokenData},
    foobar,
};
```
Comments `foo 3` and `foo 4` disappeared and there is an extra `;` after comments `bar 1` and `foo 1`.

**"Module"** granularity produce the following output:
```rust
use crate::bar; /* bar 1 */;
use crate::foo; /* foo 1 */;
use crate::foo::tokens::TokenData;
use crate::foo::{self /* foo 2 */, self};
use crate::foobar; /* foobar 1 */
use crate::{bar, foo, foobar};
```
In addition to the same issues as for "Crate", `foo::{...}` includes `self` twice: one with `foo 2` comment and one without a comment.

**"Item"** granularity produce the following output:
```rust
use crate::bar; /* bar 1 */;
use crate::foo;
use crate::foo::tokens::TokenData;
use crate::foo::{self /* foo 2 */};
use crate::foobar; /* foobar 1 */
```
Comment `foo 1` is missing because in the original code `crate::foo::self` is before `crate::foo::self /* foo 1 */`.  The `bar 1` comment does appear, since `crate::bar::self` is before `crate::foobar::self; /* foobar 1 */`.  As above, `foo 3` and `foo 4` comments disappeared and there is an extra `;` after comments `bar 1`.

Reply:
> as there are some issues, especially related to comments

wow there sure is :upside_down_face: 

We've always had comment issues with this option (which was named `merge_imports` prior to a recent rename/restructure) so I'm not too surprised by some of those, but the emission of invalid code in this type of case is particularly concerning:

```rust
use rustc_ast::ast /* abc */;
```

> I didn't add such test files as there are some issues, especially related to comments, so maybe it is better to discuss first (here or in a new issue) which of these issues should be resolved, whether there are cases where the original code should be used, what should be the expected results, etc.

Good call :+1:

Reply:
Submitted PR #4759 to fix the extra semicolon issue.

I will also try to resolve the issue where in some cases if the same import module appears twice - once with a comment and once without a comment, it appears twice in the formatted output.  E.g. in the above output for `Crate`, 'create::bar` and `crate:::foo` appear twice.